### ISSUE TITLE
Wasserstein regularization validation audio gives strange artefacts/ mixed up latents

### ISSUE THREAD

**Issue by moiseshorta** [acids-ircam/RAVE#198]
Hi,

I've been completing a couple of training experiments using the Wasserstein regularization and have noticed some strange behavior, mostly coming in later in the training stages, after Phase 1.

I'm attaching here the audio, but basically, it seems the latent from the validation/reconstruction are getting quite mixed up and causes the reconstruction to not be accurate:

https://drive.google.com/file/d/1xgy4zUM_1UWq96PkLckFwNZxpj_HsHhF/view?usp=share_link

This validation audio is from about 1.7M steps of training. I've encountered this error with two models using this same config (v2/Wasserstein)

**Comment by caillonantoine** [comment]
Hey,

I love those artifacts haha

Let me investigate :)

**Comment by moiseshorta** [comment]
@caillonantoine  hahah, yes, the artifacts are sounding good, but just thought this might be unexpected behavior. 

In my experiments, I've found that these are introduced progressively as the training goes along...

**Comment by caillonantoine** [comment]
Assuming you are working with 44.1kHz files, the problem here is that v2 has a large receptive field (~1s left, ~1s right), applied one 3s long samples. One thing that changed from v1 is that rave no longer trains on padded samples (i.e the loss is computed on a reconstruction cropped by an amount = to the receptive field). This is why you have 1s of non-sense, 1s of properly reconstructed audio, and 1s of non-sens again. However, when using the model in realtime, only the first second of reconstruction is gonna be affected, everything else will be ok :)

**Comment by moiseshorta** [comment]
Okay, that makes sense. What would fix this behavior? The dataset I'm using is one single 3hr file...

Another strange behavior I've noticed is the problem with dimension: the exported model only has 1dim, even though I am using the default 16dim configuration for Wasserstein...Also, it is not working in nn~ when exported, even though the torch training version is the same as the nn~ compiled version... I can send the exported .ts if it helps figure stuff out

**Comment by caillonantoine** [comment]
Yep, it would be helpful thanks !

**Comment by caillonantoine** [comment]
could you past the content of the `config.gin` associated with your run ?

**Comment by moiseshorta** [comment]
Sure, here it is: `from __gin__ import dynamic_registration
import cached_conv as cc
from cached_conv import convs
import rave
from rave import balancer
from rave import blocks
from rave import core
from rave import dataset
from rave import discriminator
from rave import model
from rave import pqmf
import torch
import torch.nn as nn

# Macros:
# ==============================================================================
CAPACITY = 96
DILATIONS = [1, 3, 9]
KERNEL_SIZE = 3
LATENT_SIZE = 128
N_BAND = 16
PHASE_1_DURATION = 1000000
RATIOS = [4, 4, 4, 2]
SAMPLING_RATE = 44100

# Parameters for core.AudioDistanceV1:
# ==============================================================================
core.AudioDistanceV1.log_epsilon = 1
core.AudioDistanceV1.multiscale_stft = @core.MultiScaleSTFT

# Parameters for balancer.Balancer:
# ==============================================================================
balancer.Balancer.deny_list = ['regularization']
balancer.Balancer.ema_averager = @balancer.EMA
balancer.Balancer.scale_gradients = False
balancer.Balancer.weights = {'feature_matching': 20, 'regularization': 0.1}

# Parameters for pqmf.CachedPQMF:
# ==============================================================================
pqmf.CachedPQMF.attenuation = 100
pqmf.CachedPQMF.n_band = %N_BAND

# Parameters for discriminator.CombineDiscriminators:
# ==============================================================================
discriminator.CombineDiscriminators.discriminators = \
    [@discriminator.MultiPeriodDiscriminator,
     @discriminator.MultiScaleDiscriminator]

# Parameters for cc.Conv1d:
# ==============================================================================
cc.Conv1d.bias = False

# Parameters for variational/cc.Conv1d:
# ==============================================================================
variational/cc.Conv1d.bias = False

# Parameters for scales/torch.nn.Conv1d:
# ==============================================================================
scales/torch.nn.Conv1d.bias = True
scales/torch.nn.Conv1d.device = None
scales/torch.nn.Conv1d.dilation = 1
scales/torch.nn.Conv1d.dtype = None
scales/torch.nn.Conv1d.groups = 1
scales/torch.nn.Conv1d.padding = 0
scales/torch.nn.Conv1d.padding_mode = 'zeros'
scales/torch.nn.Conv1d.stride = 1

# Parameters for periods/nn.Conv2d:
# ==============================================================================
periods/nn.Conv2d.bias = True
periods/nn.Conv2d.device = None
periods/nn.Conv2d.dilation = 1
periods/nn.Conv2d.dtype = None
periods/nn.Conv2d.groups = 1
periods/nn.Conv2d.padding = 0
periods/nn.Conv2d.padding_mode = 'zeros'
periods/nn.Conv2d.stride = 1

# Parameters for periods/discriminator.ConvNet:
# ==============================================================================
periods/discriminator.ConvNet.capacity = %CAPACITY
periods/discriminator.ConvNet.conv = @nn.Conv2d
periods/discriminator.ConvNet.in_size = 1
periods/discriminator.ConvNet.kernel_size = (5, 1)
periods/discriminator.ConvNet.n_layers = 4
periods/discriminator.ConvNet.out_size = 1
periods/discriminator.ConvNet.stride = 4

# Parameters for scales/discriminator.ConvNet:
# ==============================================================================
scales/discriminator.ConvNet.capacity = %CAPACITY
scales/discriminator.ConvNet.conv = @torch.nn.Conv1d
scales/discriminator.ConvNet.in_size = 1
scales/discriminator.ConvNet.kernel_size = 15
scales/discriminator.ConvNet.n_layers = 4
scales/discriminator.ConvNet.out_size = 1
scales/discriminator.ConvNet.stride = 4

# Parameters for cc.ConvTranspose1d:
# ==============================================================================
cc.ConvTranspose1d.bias = False

# Parameters for balancer.EMA:
# ==============================================================================
balancer.EMA.beta = 0.999

# Parameters for variational/blocks.EncoderV2:
# ==============================================================================
variational/blocks.EncoderV2.capacity = %CAPACITY
variational/blocks.EncoderV2.data_size = %N_BAND
variational/blocks.EncoderV2.dilations = %DILATIONS
variational/blocks.EncoderV2.keep_dim = False
variational/blocks.EncoderV2.kernel_size = %KERNEL_SIZE
variational/blocks.EncoderV2.latent_size = %LATENT_SIZE
variational/blocks.EncoderV2.n_out = 2
variational/blocks.EncoderV2.ratios = %RATIOS
variational/blocks.EncoderV2.recurrent_layer = None

# Parameters for blocks.GeneratorV2:
# ==============================================================================
blocks.GeneratorV2.amplitude_modulation = True
blocks.GeneratorV2.capacity = %CAPACITY
blocks.GeneratorV2.data_size = %N_BAND
blocks.GeneratorV2.dilations = %DILATIONS
blocks.GeneratorV2.keep_dim = False
blocks.GeneratorV2.kernel_size = %KERNEL_SIZE
blocks.GeneratorV2.latent_size = %LATENT_SIZE
blocks.GeneratorV2.ratios = %RATIOS
blocks.GeneratorV2.recurrent_layer = None

# Parameters for convs.get_padding:
# ==============================================================================
convs.get_padding.dilation = 1
convs.get_padding.mode = 'centered'
convs.get_padding.stride = 1

# Parameters for periods/convs.get_padding:
# ==============================================================================
periods/convs.get_padding.dilation = 1

# Parameters for scales/convs.get_padding:
# ==============================================================================
scales/convs.get_padding.dilation = 1

# Parameters for variational/convs.get_padding:
# ==============================================================================
variational/convs.get_padding.dilation = 1
variational/convs.get_padding.mode = 'centered'
variational/convs.get_padding.stride = 1

# Parameters for discriminator.MultiPeriodDiscriminator:
# ==============================================================================
discriminator.MultiPeriodDiscriminator.convnet = @periods/discriminator.ConvNet
discriminator.MultiPeriodDiscriminator.periods = [2, 3, 5, 7, 11]

# Parameters for discriminator.MultiScaleDiscriminator:
# ==============================================================================
discriminator.MultiScaleDiscriminator.convnet = @scales/discriminator.ConvNet
discriminator.MultiScaleDiscriminator.n_discriminators = 3

# Parameters for core.MultiScaleSTFT:
# ==============================================================================
core.MultiScaleSTFT.magnitude = True
core.MultiScaleSTFT.normalized = False
core.MultiScaleSTFT.num_mels = None
core.MultiScaleSTFT.sample_rate = %SAMPLING_RATE
core.MultiScaleSTFT.scales = [2048, 1024, 512, 256, 128]

# Parameters for blocks.normalization:
# ==============================================================================
blocks.normalization.mode = 'weight_norm'

# Parameters for periods/blocks.normalization:
# ==============================================================================
periods/blocks.normalization.mode = 'weight_norm'

# Parameters for scales/blocks.normalization:
# ==============================================================================
scales/blocks.normalization.mode = 'weight_norm'

# Parameters for variational/blocks.normalization:
# ==============================================================================
variational/blocks.normalization.mode = 'weight_norm'

# Parameters for model.RAVE:
# ==============================================================================
model.RAVE.audio_distance = @core.AudioDistanceV1
model.RAVE.balancer = @balancer.Balancer
model.RAVE.decoder = @blocks.GeneratorV2
model.RAVE.discriminator = @discriminator.CombineDiscriminators
model.RAVE.encoder = @blocks.VariationalEncoder
model.RAVE.feature_matching_fun = @feature_matching/core.mean_difference
model.RAVE.gan_loss = @core.hinge_gan
model.RAVE.latent_size = %LATENT_SIZE
model.RAVE.multiband_audio_distance = @core.AudioDistanceV1
model.RAVE.num_skipped_features = 1
model.RAVE.phase_1_duration = %PHASE_1_DURATION
model.RAVE.pqmf = @pqmf.CachedPQMF
model.RAVE.sampling_rate = %SAMPLING_RATE
model.RAVE.update_discriminator_every = 4
model.RAVE.valid_signal_crop = True
model.RAVE.warmup_quantize = None

# Parameters for dataset.split_dataset:
# ==============================================================================
dataset.split_dataset.max_residual = 1000

# Parameters for blocks.VariationalEncoder:
# ==============================================================================
blocks.VariationalEncoder.encoder = @variational/blocks.EncoderV2`

**Comment by caillonantoine** [comment]
This is not a wasserstein model, but a variational one. Since you are training on a very (very) small dataset, your validation set must be even smaller, thus the latent space dimensionality estimation might get wrongly estimated. I suggest you re-export your model with `--fidelity 0.999999` or something like this

**Comment by moiseshorta** [comment]
Hmm, that's very strange, I might have resumed the model without the --config v2 parameter set. Thanks for pointing that out. Would a 5hr dataset be considered small?
